README.md
# [MyENV: Environment Variable Parsing using Type annotations][repo_ref]

MyENV parses you're environment variables using type annotations.
This allows you to configure your app/service as layed out by
[12factor.net/config](https://12factor.net/config), while keeping
your code type safe.

Project/Repo:

[![MIT License][license_img]][license_ref]
[![Supported Python Versions][pyversions_img]][pyversions_ref]
[![PyCalVer v201902.0006][version_img]][version_ref]
[![PyPI Version][pypi_img]][pypi_ref]
[![PyPI Downloads][downloads_img]][downloads_ref]

Code Quality/CI:

[![Build Status][build_img]][build_ref]
[![Type Checked with mypy][mypy_img]][mypy_ref]
[![Code Coverage][codecov_img]][codecov_ref]
[![Code Style: sjfmt][style_img]][style_ref]


|                 Name                |    role           |  since  | until |
|-------------------------------------|-------------------|---------|-------|
| Manuel Barkhau (mbarkhau@gmail.com) | author/maintainer | 2018-09 | -     |


<!--
  To update the TOC:
  $ pip install md-toc
  $ md_toc -i gitlab README.md
-->


[](TOC)

- [Environment Variables and Configuration](#environment-variables-and-configuration)
  - [Declaration](#declaration)
  - [Parsing](#parsing)
- [my_service/cfg.py](#my_service-cfg-py)
  - [Config Files](#config-files)
- [config/prod.env](#config-prod-env)

[](TOC)

## Environment Variables and Configuration

In order of priority, configuration is parsed from

 1. Environment variables
 2. Configuration files
 3. Defaults defined in source code


## Declaration

The `myenv` module provides a convenient way to do this parsing.
As far as your application code is concerned, it is enough to
declare a subclass of `myenv.BaseEnv`. Instances of this
subclass are populated from config files and the environment.

```python
import myenv


class Database(myenv.BaseEnv):

    _environ_prefix = "DATABASE_"

    vendor    : str  = "postgres"
    host      : str  = "127.0.0.1"
    port      : int  = 5432
    user      : str  = "myuser"
    password  : str
    name      : str  = "app_db_v1"

    @property
    def url(self) -> str:
        db = self
        return f"{db.vendor}://{db.user}:{db.password}@{db.host}:{db.port}/{db.name}"
```

For each annoatated member of `DBEnv` declare the 1. name, 2. type and
3. an optional default  variable. Members without a
default must be set by an environment variable or configuration
file, otherwise a `KeyError` will be raised.


## Parsing

To use the above configuration class, simply instanciate it.

```python
# my_service/cfg.py
import sqlalchemay as sqla

db_cfg = Database()     # populated from os.environ
db_cfg.port             # 12345          (parsed from DATABASE_PORT)
db_cfg.password         # "supersecret"  (parsed from DATABASE_PASSWORD)
db_cfg.url              # "mysql://myuser:supersecret@127.0.0.1:12345/mydb"

engine = sqla.create_engine(db_cfg.url)
```

In case you're worried about the instantiation of `Database()`,
it will return a singleton instance, so the `os.environ` and
configuration files are parsed only once.


## Config Files

When parsing configs, the following paths are parsed if they
exist. By default the configs are loaded from `${PWD}/config/`,
but you can override the location of configuration files by
setting the environment variable `ENV_CONFIG_DIR`.

 - `${ENV_CONFIG_DIR}/${ENV}.env`
 - `${ENV_CONFIG_DIR}/prod.env`

Any variables defined in these files will be set, unless it was
already set in the environ or a previous config file.

This approach allows you to satisfy typical use cases in which a
service is hosted:

 1. Development Machine
 2. Stage/Production Environment

You can put use case specifc configuration files in your project
source tree, such as:

 - `project/config/dev.env`
 - `project/config/stage.env`
 - `project/config/prod.env`

To parse the appropriate config file, all you have to do is set a
single environment variable `ENV=<envname>`. If `ENV` is not set,
it will fall back to `ENV=prod`.

The `*.env` config files are flat text files, with one KEY=value
mapping per line. The only lines which are parsed, are lines which
begin with an all upper case token, followed by an `=` character:

```ini
# config/prod.env
DATABASE_PORT=12345
DATABASE_USER=prod_user
DATABASE_NAME=prod_db
DATABASE_PASSWORD=supersecret
```

For sensitive configuration parameters, such as passwords and
authentication tokens, you may prefer to write config files
outside of your source tree, or to provide them always and only
via environment variables.



[repo_ref]: https://gitlab.com/mbarkhau/myenv

[build_img]: https://gitlab.com/mbarkhau/myenv/badges/master/pipeline.svg
[build_ref]: https://gitlab.com/mbarkhau/myenv/pipelines

[codecov_img]: https://gitlab.com/mbarkhau/myenv/badges/master/coverage.svg
[codecov_ref]: https://mbarkhau.gitlab.io/myenv/cov

[license_img]: https://img.shields.io/badge/License-MIT-blue.svg
[license_ref]: https://gitlab.com/mbarkhau/myenv/blob/master/LICENSE

[mypy_img]: https://img.shields.io/badge/mypy-checked-green.svg
[mypy_ref]: https://mbarkhau.gitlab.io/myenv/mypycov

[style_img]: https://img.shields.io/badge/code%20style-%20sjfmt-f71.svg
[style_ref]: https://gitlab.com/mbarkhau/straitjacket/

[pypi_img]: https://img.shields.io/badge/PyPI-wheels-green.svg
[pypi_ref]: https://pypi.org/project/myenv/#files

[downloads_img]: https://pepy.tech/badge/myenv/month
[downloads_ref]: https://pepy.tech/project/myenv

[version_img]: https://img.shields.io/static/v1.svg?label=PyCalVer&message=v201902.0006&color=blue
[version_ref]: https://pypi.org/project/pycalver/

[pyversions_img]: https://img.shields.io/pypi/pyversions/myenv.svg
[pyversions_ref]: https://pypi.python.org/pypi/myenv



CHANGELOG.md
# Changelog for https://gitlab.com/mbarkhau/myenv

## v201907.0008

 - Add __repr__


## v201902.0007

 - Packaging updates
 - Better test and mypy coverage


## v201812.0005-beta

 - Add config parsing
 - Simplify instantiation (no need to import myenv to use subclasses)
 - Switch to https://gitlab.com/mbarkhau/myenv
 - Use https://gitlab.com/mbarkhau/bootstrapit


## v201809.0001-beta

 - Initial release


test/test_myenv.py
import os
import myenv
import pathlib as pl
import typing as typ


# dbenv_cls and testenv_cls are declared to avoid PytestWarning:
#   cannot collect test class 'TestEnv' because it has a __init__ constructor


def dbenv_cls():
    class DBEnv(myenv.BaseEnv):

        _environ_prefix = "MYAPP_DB_"

        host     : str = "127.0.0.1"
        port     : int = 5432
        name     : str = "app_db_v1"
        user     : str = "app_user"
        password : str
        read_only: bool = True
        ddl      : pl.Path = (pl.Path(".") / "create_tables.sql")

        @property
        def url(self) -> str:
            db = self
            return f"postgres://{db.user}:{db.password}@{db.host}:{db.port}/{db.name}"

    return DBEnv


def dbenv_environ():
    return {
        'MYAPP_DB_HOST'     : "1.2.3.4",
        'MYAPP_DB_PORT'     : "1234",
        'MYAPP_DB_NAME'     : "app_db_v2",
        'MYAPP_DB_USER'     : "new_user",
        'MYAPP_DB_PASSWORD' : "secret",
        'MYAPP_DB_READ_ONLY': "0",
        'MYAPP_DB_DDL'      : "~/mkdb.sql",
    }


def testenv_cls():
    class TestEnv(myenv.BaseEnv):

        str_val_wo_default  : str
        str_val             : str = "foo"
        int_val_wo_default  : int
        int_val             : int = 123
        bool_val_wo_default : bool
        bool_val            : bool = True
        float_val_wo_default: float
        float_val           : float = 12.34
        strs_val_wo_default : typ.List[str]
        strs_val            : typ.List[str] = ['bar']
        ints_val_wo_default : typ.Set[int]
        ints_val            : typ.Set[int] = {1, 2}
        path_val_wo_default : pl.Path
        path_val            : pl.Path = pl.Path("file.txt")
        paths_val_wo_default: typ.List[pl.Path]
        paths_val           : typ.List[pl.Path] = [
            pl.Path("file1.txt"),
            pl.Path("file2.txt"),
            pl.Path("file3.txt"),
        ]
        custom_wo_default: lambda val: "".join(reversed(val))

    return TestEnv


def test_parse():
    environ        = dbenv_environ()
    environ_before = str(os.environ)

    dbenv = dbenv_cls()(environ=environ)

    environ_after = str(os.environ)
    assert environ_before == environ_after

    assert set(dbenv._varnames()) == set(environ.keys())

    assert dbenv.host     == "1.2.3.4"
    assert dbenv.port     == 1234
    assert dbenv.name     == "app_db_v2"
    assert dbenv.user     == "new_user"
    assert dbenv.password == "secret"
    assert dbenv.read_only is False
    assert isinstance(dbenv.ddl, pl.Path)
    assert str(dbenv.ddl).endswith("mkdb.sql")

    assert dbenv.url == "postgres://new_user:secret@1.2.3.4:1234/app_db_v2"

    attrnames = [attrname for attrname in dbenv.__annotations__ if not attrname.startswith("_")]
    assert len(attrnames) == 7


def test_defaults():
    environ_before = str(os.environ)

    dbenv = dbenv_cls()(environ={'MYAPP_DB_PASSWORD': "secret"})

    environ_after = str(os.environ)
    assert environ_before == environ_after

    assert dbenv.host     == "127.0.0.1"
    assert dbenv.port     == 5432
    assert dbenv.name     == "app_db_v1"
    assert dbenv.user     == "app_user"
    assert dbenv.password == "secret"
    assert dbenv.read_only is True

    assert isinstance(dbenv.ddl, pl.Path)
    assert str(dbenv.ddl).endswith("create_tables.sql")

    attrnames = [attrname for attrname in dbenv.__annotations__ if not attrname.startswith("_")]
    assert len(attrnames) == 7


def test_errors():
    try:
        dbenv = testenv_cls()(environ={})
        assert False
    except KeyError as err:
        assert "STR_VAL_WO_DEFAULT" in str(err)

    try:
        dbenv = testenv_cls()(environ={'STR_VAL_WO_DEFAULT': "bar"})
        assert False
    except KeyError as err:
        assert "INT_VAL_WO_DEFAULT" in str(err)

    good_environ = {
        'STR_VAL_WO_DEFAULT'  : "bar",
        'INT_VAL_WO_DEFAULT'  : "123",
        'BOOL_VAL_WO_DEFAULT' : "TRUE",
        'FLOAT_VAL_WO_DEFAULT': "123.456",
        'STRS_VAL_WO_DEFAULT' : "baz:buz",
        'INTS_VAL_WO_DEFAULT' : "7:89",
        'PATH_VAL_WO_DEFAULT' : "fileA.txt",
        'PATHS_VAL_WO_DEFAULT': "fileA.txt:fileB.txt:fileC.txt",
        'CUSTOM_WO_DEFAULT'   : "abcd",
    }
    dbenv = testenv_cls()(environ=good_environ)

    assert dbenv.str_val             == "foo"
    assert dbenv.str_val_wo_default  == "bar"
    assert dbenv.strs_val            == ['bar']
    assert dbenv.strs_val_wo_default == ["baz", 'buz']
    assert dbenv.ints_val            == {1, 2}
    assert dbenv.ints_val_wo_default == {7, 89}
    assert dbenv.custom_wo_default   == "dcba"

    try:
        bad_environ = good_environ.copy()
        bad_environ['INT_VAL'] = "abc"
        dbenv = testenv_cls()(environ=bad_environ)
        assert False
    except ValueError as err:
        assert 'INT_VAL' in str(err)
        assert 'abc' in str(err)
        assert "TestEnv.int_val" in str(err)

    try:
        bad_environ = good_environ.copy()
        bad_environ['INT_VAL'] = "123af"
        dbenv = testenv_cls()(environ=bad_environ)
        assert False
    except ValueError as err:
        assert 'INT_VAL' in str(err)
        assert '123af' in str(err)
        assert "TestEnv.int_val" in str(err)

    try:
        bad_environ = good_environ.copy()
        bad_environ['BOOL_VAL'] = "yes"
        dbenv = testenv_cls()(environ=bad_environ)
        assert False
    except ValueError as err:
        assert 'BOOL_VAL' in str(err)
        assert 'yes' in str(err)
        assert "TestEnv.bool_val" in str(err)

    try:
        bad_environ = good_environ.copy()
        bad_environ['FLOAT_VAL'] = "abc.23"
        dbenv = testenv_cls()(environ=bad_environ)
        assert False
    except ValueError as err:
        assert 'FLOAT_VAL' in str(err)
        assert "abc.23" in str(err)
        assert "TestEnv.float_val" in str(err)


def test_self_test():
    myenv.__self_test()


def test_asdict():
    environ = dbenv_environ()
    dbenv   = dbenv_cls()(environ=environ)

    d = dbenv._asdict()

    assert d['host'] == "1.2.3.4"
    assert d['port'] == 1234
    assert d['name'] == "app_db_v2"
    assert d['user'] == "new_user"
    assert d['password'] == "secret"
    assert d['ddl'] == pl.Path("~/mkdb.sql")
    assert not d['read_only']


def test_repr():
    repr_vals = [
        "host=" + repr("1.2.3.4"),
        "port=" + repr(1234),
        "name=" + repr("app_db_v2"),
        "user=" + repr("new_user"),
        "password=" + repr("secret"),
        "ddl=" + repr(pl.Path("~/mkdb.sql")),
        "read_only=False",
    ]

    environ = dbenv_environ()
    dbenv   = dbenv_cls()(environ=environ)

    dbenv_repr = repr(dbenv)
    assert dbenv_repr.startswith("DBEnv(")

    for val in repr_vals:
        assert val + "," in dbenv_repr or ", " + val + ")" in dbenv_repr


bootstrapit.sh
#!/bin/bash
# Bootstrapit Project Configuration

AUTHOR_NAME="Manuel Barkhau"
AUTHOR_EMAIL="mbarkhau@gmail.com"

KEYWORDS="environ variables mypy config configuration"
DESCRIPTION="Type safe environment variables parsing."


LICENSE_ID="MIT"

PACKAGE_NAME="myenv"
MODULE_NAME="myenv"
GIT_REPO_NAMESPACE="mbarkhau"
GIT_REPO_DOMAIN="gitlab.com"

DEFAULT_PYTHON_VERSION="python=3.6"
SUPPORTED_PYTHON_VERSIONS="python=3.6 python=3.7"

PACKAGE_VERSION="v201902.0007"

# GIT_REPO_URL=https://${GIT_REPO_DOMAIN}/${GIT_REPO_NAMESPACE}/${PACKAGE_NAME}


IS_PUBLIC=1


## Download and run the actual update script

if [[ $KEYWORDS == "keywords used on pypi" ]]; then
    echo "Default bootstrapit config detected.";
    echo "Did you forget to update parameters in your 'bootstrapit.sh' ?"
    exit 1;
fi

PROJECT_DIR=$(dirname "$0");

if ! [[ -f "$PROJECT_DIR/scripts/bootstrapit_update.sh" ]]; then
    mkdir -p "$PROJECT_DIR/scripts/";
    RAW_FILES_URL="https://gitlab.com/mbarkhau/bootstrapit/raw/master";
    curl --silent "$RAW_FILES_URL/scripts/bootstrapit_update.sh" \
        > "$PROJECT_DIR/scripts/bootstrapit_update.sh.tmp";
    mv "$PROJECT_DIR/scripts/bootstrapit_update.sh.tmp" \
        "$PROJECT_DIR/scripts/bootstrapit_update.sh";
fi

source "$PROJECT_DIR/scripts/bootstrapit_update.sh";


src/myenv/__init__.py
# This file is part of the myenv project
# https://gitlab.com/mbarkhau/myenv
#
# Copyright (c) 2019 Manuel Barkhau (mbarkhau@gmail.com) - MIT License
# SPDX-License-Identifier: MIT
"""Environment variable parsing using type annotations.

NOTE: Normally you don't need to declare environ, since the
default is to just use os.environ. For this doctest however, we
don't want to manipulate the os.environ and so we create a mock
instead.

Usage:

>>> mock_environ = {
...     'CREDENTIALS_USER': "franz",
...     'CREDENTIALS_KEY': "supersecret",
... }
>>> class Credentials(BaseEnv):
...     _environ_prefix = "CREDENTIALS_"
...     user : str = "user"
...     key : str
...
>>> creds = Credentials(environ=mock_environ)
>>> creds == Credentials(user='franz', key='supersecret')
True
>>> creds.user
'franz'
>>> creds.key
'supersecret'
>>> creds._varnames()
['CREDENTIALS_USER', 'CREDENTIALS_KEY']
>>> creds._asdict()
{'user': 'franz', 'key': 'supersecret'}
>>> Credentials(environ=mock_environ) is Credentials(environ=mock_environ)
True
"""

import os
import pathlib as pl
import typing as typ
import typing_inspect


__version__ = "v201902.0007"


Environ = typ.MutableMapping[str, str]


EnvType = typ.TypeVar('EnvType', bound='BaseEnv')


def _iter_env_config(env_name: str) -> typ.Iterable[typ.Tuple[str, str]]:
    env_config_dir = os.getenv('ENV_CONFIG_DIR', ".")
    CONFIG_DIR     = pl.Path(env_config_dir) / "config"
    config_files   = [CONFIG_DIR / (env_name + ".env")]
    if env_name != 'prod':
        config_files.append(CONFIG_DIR / "prod.env")

    for config_file in config_files:
        if not config_file.exists():
            continue

        fh: typ.IO[str]

        with config_file.open(mode="rt", encoding="utf-8") as fh:
            config_lines = fh.readlines()

        for line in config_lines:
            if "=" in line and not line.startswith("#"):
                name, value = line.strip().split("=", 1)
                yield name.strip(), value


_environ_initialized: typ.Set[int] = set()


def _init_environ(environ: Environ = None) -> None:
    if id(environ) in _environ_initialized:
        return

    if environ is None:
        environ = os.environ

    env_name = os.getenv('ENV', 'prod').lower()

    for name, value in _iter_env_config(env_name):
        if name not in environ:
            environ[name] = value.strip()

    _environ_initialized.add(id(environ))


__fallback_sentinel__ = '__fallback_sentinel__'

# NOTE (2018-11-30 mb): I couldn't find out how to express
#   that a parameter one of a few type itself, rather than
#   an instance of one of a few types.
FieldType = typ.Any

# FieldValue = typ.Union[str, int, float, bool, pl.Path]
FieldValue     = typ.Any
ListFieldValue = typ.List[typ.Any]


def _parse_bool(val: str) -> bool:
    if val.lower() in ("1", "true"):
        return True
    elif val.lower() in ("0", "false"):
        return False
    else:
        raise ValueError(val)


def _parse_list_val(val: str, ftype: FieldType) -> ListFieldValue:
    maybe_args: typ.Optional[typ.Tuple[type]] = getattr(ftype, '__args__', None)
    if maybe_args is None:
        raise TypeError(ftype)

    if len(maybe_args) == 1:
        member_type: type = maybe_args[0]
    else:
        raise TypeError(ftype)

    list_strvals = [strval for strval in val.split(os.pathsep)]

    if issubclass(member_type, str):
        return list_strvals
    elif issubclass(member_type, pl.Path):
        return [pl.Path(listval) for listval in list_strvals]
    elif issubclass(member_type, int):
        return [int(listval, 10) for listval in list_strvals]
    elif issubclass(member_type, float):
        return [float(listval) for listval in list_strvals]
    elif issubclass(member_type, bool):
        return [_parse_bool(listval) for listval in list_strvals]
    else:
        raise TypeError(ftype)


def _parse_val(val: str, ftype: type) -> FieldValue:
    ftype_type = type(ftype)
    if ftype_type == type:
        if issubclass(ftype, str):
            return val
        elif issubclass(ftype, bool):
            return _parse_bool(val)
        elif issubclass(ftype, int):
            return int(val, 10)
        elif issubclass(ftype, float):
            return float(val)
        elif issubclass(ftype, pl.Path):
            return pl.Path(val)
        else:
            raise TypeError(ftype)
    elif typing_inspect.get_origin(ftype) in (list, typ.List):
        return _parse_list_val(val, ftype)
    elif typing_inspect.get_origin(ftype) in (set, typ.Set):
        return set(_parse_list_val(val, ftype))
    elif callable(ftype):
        return ftype(val)
    else:
        raise TypeError(ftype)


# Cache for already loaded environment configs. Environment
# variables are only parsed once during initialization.

EnvMapKey = typ.Tuple[typ.Type[EnvType], int]
EnvMap    = typ.Dict[EnvMapKey, EnvType]

_envmap: EnvMap = {}


# NOTE (2019-02-24 mb): Could never get this to work,
#   I thought it was more important to spend time on
#   test coverage.
# EnvironKWArgs = mypyext.TypedDict('EnvironKWArgs', {'environ': Environ})


class _Singleton(type):
    def __call__(cls, *args, **kwargs) -> EnvType:
        env_cls = typ.cast(typ.Type[EnvType], cls)

        environ: Environ
        if 'environ' in kwargs:
            environ = kwargs['environ']
        elif not kwargs:
            environ = os.environ
        else:
            # init with kwargs (not via environ)
            return env_cls.__new__(env_cls, *args, **kwargs)

        envmap_key = (env_cls, id(environ))

        if envmap_key not in _envmap:
            _init_environ(environ)
            _envmap[envmap_key] = env_cls.__new__(env_cls, environ=environ)

        return _envmap[envmap_key]


class _Field(typ.NamedTuple):

    fname   : str
    ftyp    : FieldType
    env_key : str
    fallback: FieldValue


InitKWArgs = typ.MutableMapping[str, typ.Any]


class BaseEnv(metaclass=_Singleton):
    """The main Base class.

    Subclasses of BaseEnv are only instantiated once (singleton).
    """

    _environ_prefix: typ.Optional[str] = None

    @classmethod
    def _iter_fields(cls) -> typ.Iterable[_Field]:
        prefix = cls._environ_prefix or ""
        for fname, ftyp in cls.__annotations__.items():
            fallback = getattr(cls, fname, __fallback_sentinel__)
            env_key  = (prefix + fname).upper()
            yield _Field(fname, ftyp, env_key, fallback)

    @classmethod
    def _update_kwargs_from_environ(cls, environ: Environ) -> InitKWArgs:
        typename = cls.__name__
        init_kwargs: InitKWArgs = {}

        for field in cls._iter_fields():
            if field.env_key in environ:
                try:
                    raw_env_val = environ[field.env_key]
                    init_kwargs[field.fname] = _parse_val(raw_env_val, field.ftyp)
                except ValueError as err:
                    raise ValueError(
                        f"Invalid value '{raw_env_val}' for {field.env_key}. "
                        f"Attepmted to parse '{typename}.{field.fname}' with '{field.ftyp}'.",
                        err,
                    )
            elif field.fallback != __fallback_sentinel__:
                init_kwargs[field.fname] = field.fallback
            else:
                raise KeyError(
                    f"No environment variable {field.env_key} "
                    + f"found for field {typename}.{field.fname}"
                )

        return init_kwargs

    def __new__(cls, *args, **kwargs) -> EnvType:
        """Create a new env instance.

        This should not be called from outside of myenv.
        """
        init_kwargs: InitKWArgs

        if 'environ' in kwargs:
            init_kwargs = cls._update_kwargs_from_environ(kwargs['environ'])
        else:
            init_kwargs = kwargs

        env = super(BaseEnv, cls).__new__(cls)
        env.__init__(*args, **init_kwargs)
        return env

    def __init__(self, *args, **kwargs) -> None:
        for key, val in kwargs.items():
            setattr(self, key, val)

    def _varnames(self) -> typ.List[str]:
        """Create list with names as they are read from os.environ and configs.

        >>> creds = Credentials(user='franz', key='supersecret')
        >>> creds._varnames()
        ['CREDENTIALS_USER', 'CREDENTIALS_KEY']
        """
        prefix = self._environ_prefix or ""
        return [
            (prefix + attrname).upper()
            for attrname in type(self).__annotations__
            if not attrname.startswith("_")
        ]

    def _asdict(self) -> typ.Dict[str, typ.Any]:
        """Create a dict populated with keys/values from annotated fields.

        >>> creds = Credentials(user='franz', key='supersecret')
        >>> creds._asdict()
        {'user': 'franz', 'key': 'supersecret'}
        """
        return {
            field.fname: getattr(self, field.fname)
            for field in self._iter_fields()
            if not field.fname.startswith("_")
        }

    def __eq__(self, other: object) -> bool:
        """Deep equality check for all annotated fields."""
        return isinstance(other, BaseEnv) and self._asdict() == other._asdict()

    def __repr__(self) -> str:
        """Return repr(self)."""
        kwargs: typ.List[str] = []
        for field in self._iter_fields():
            if not field.fname.startswith("_"):
                field_val_str = repr(getattr(self, field.fname))
                kwarg         = f"{field.fname}={field_val_str}"
                kwargs.append(kwarg)

        kwarg_str = ", ".join(kwargs)
        typename  = self.__class__.__name__
        return f"{typename}({kwarg_str})"


def parse(env_type: typ.Type[EnvType], environ: Environ = os.environ) -> EnvType:
    """Create an instance of an env.

    This is depricated, just instantiate like a normal class
    - myenv.parse(MyEnv)
    + MyEnv()
    """
    return env_type(environ=environ)


class Credentials(BaseEnv):
    """Helper class used in doc tests."""

    _environ_prefix: str = "CREDENTIALS_"
    user           : str = "user"
    key            : str


def __self_test():
    """Some code for mypy to type check.

    Since the unittests are not type checked, this code is to make sure
    that access to the properties of a subclass of BaseEnv are indeed
    detected to be of the declared type. Considering that these fields are
    always populated only after runtime type checks (which are opaque to
    mypy) this bit of code demonstrates that the whole purpose of this
    library is being satisfied.
    """

    class _TestEnv(BaseEnv):
        str_val  : str
        int_val  : int
        bool_val : bool
        float_val: float
        strs_val : typ.List[str]
        ints_val : typ.Set[int]
        path_val : pl.Path
        paths_val: typ.List[pl.Path]

    environ: Environ = {
        'STR_VAL'  : "bar",
        'INT_VAL'  : "123",
        'BOOL_VAL' : "TRUE",
        'FLOAT_VAL': "123.456",
        'STRS_VAL' : "baz:buz",
        'INTS_VAL' : "7:89",
        'PATH_VAL' : "fileA.txt",
        'PATHS_VAL': "fileA.txt:fileB.txt:fileC.txt",
    }

    testenv = _TestEnv(environ=environ)
    str_val: str = testenv.str_val
    assert str_val == "bar"
    int_val: int = testenv.int_val
    assert int_val == 123
    bool_val: bool = testenv.bool_val
    assert bool_val
    float_val: float = testenv.float_val
    assert float_val == 123.456
    strs_val: typ.List[str] = testenv.strs_val
    assert strs_val == ["baz", "buz"]
    ints_val: typ.Set[int] = testenv.ints_val
    assert ints_val == {7, 89}
    path_val: pl.Path = testenv.path_val
    assert path_val == pl.Path("fileA.txt")
    paths_val: typ.List[pl.Path] = testenv.paths_val
    assert paths_val == [pl.Path("fileA.txt"), pl.Path("fileB.txt"), pl.Path("fileC.txt")]
